---
title: "The TREGEL user guide"
author: "Sven Berres"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
abstract: >
  TREGEL calculates overlap between user defined genetic region datasets. Any regions can be supplied i.e. genes, SNPs, or reads from sequencing experiments. Key numbers help analyse the extend of overlaps which can also be visualized at a genomic level. TREGEL package version: `r packageVersion("TREGEL")`

  
output:
    rmarkdown::html_document:
    highlight: pygments
    toc: true
    fig_width: 5

vignette: >
  %\VignetteIndexEntry{TREGEL3}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
  %\VignetteBuilder{knitr}
---

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy = FALSE,
                      cache = FALSE,
                      dev = "png",
                      message = FALSE, error = FALSE, warning = TRUE)
if(dir.exists("../inst/extdata")){
  knitr::opts_knit$set(root.dir = "../inst/extdata")
}else{
  knitr::opts_knit$set(root.dir = "../extdata")
}
```	

## Introduction

**Note:** if you use TREGEL in published research, please cite:

#> "Sven Berres" (2021)
#> *journalX*, **Jnumber**:pageNr.
#> [doiX](http://linkX)
Other Bioconductor packages with similar aims are
[PackageX](http://linkX/linkX).

## Setup

Load the TREGEL package: 

```{r message=TRUE, warning=FALSE}
library(TREGEL)

```
## Quick start

To start up TREGEL you have to generate a `TREGELDataSet` that is used to store your data sets and additional information about the analysis that you are conducting. Query and subjects files can be conveniently stored in their own folders as GRanges objects in form of stored .rds / .RDS files. We point TREGEL to the correct location by supplying a path for each folder with the character vectors `queryFolder` and `subjectFolder`. In this vignette we are using lightweight query and subject example data sets to show TREGEL's functionality.

```{r InitializeTREGEL, message=TRUE, echo=TRUE,warning=FALSE}
myQueryFolder <- file.path(system.file('extdata', package = 'TREGEL'),"query")
mySubjectFolder <- file.path(system.file('extdata', package = 'TREGEL'),"subject")

myTREGEL <- TREGELDataSetFromDir(queryFolder=myQueryFolder,subjectFolder=mySubjectFolder)
```
By monitoring TREGEL's metadata information you can make sure the input paths you supplied are stored correctly.
```{r checkMetadata, message=TRUE, echo=TRUE,warning=TRUE}
S4Vectors::metadata(myTREGEL)
```
Query and subject datasets are read by `loadAnnotations()` and stored in the `TREGELDataSet` as `GRanges` objects.
We are going to read in the following example datasets:

* query "genes" (4 regions) 
* subject "CGI" (3 CpG islands)
* subject "TFBS" (7 Transcription factor binding sites) 
```{r loadAnnotations, message=TRUE, echo=TRUE,warning=TRUE}
myTREGEL <- loadAnnotations(myTREGEL)
myTREGEL
```
TREGEL uses your dataset file names to label query and subjects internally, we can check these names by using the `names()` function since every `TREGELDataSet` is a `GRangesList`. 
```{r checkDatasetNames, message=TRUE, echo=TRUE,warning=TRUE}
names(myTREGEL)
```
Next, we find overlaps between query and subject(s) datasets by using `fOverlaps()`. Any existing overlap is listed in `detailDT` and stored as a `data.table`.
```{r fOverlaps, message=TRUE, echo=TRUE,warning=TRUE}
myTREGEL <- fOverlaps(myTREGEL)
head(S4Vectors::metadata(myTREGEL)$detailDT)
```


The summary plot shows a useful information of the number of overlaps between your datasets. 
```{r sumPlot, echo=TRUE, message=TRUE, warning=TRUE,fig.dim = c(30/2.54, 20/2.54)}
 myTREGEL <- sumPlot(myTREGEL)
 S4Vectors::metadata(myTREGEL)$barplot_summary
```
Using the `Gviz` visualization each query can be displayed with all overlapping subject elements. Choose labels for all region tracks by supplying a `trackRegionLabels` vector. Plots are stored in the same location as your dataset files.
```{r gvizPlot, message=TRUE, echo=TRUE,warning=TRUE}
 myTREGEL <- gvizPlot(myTREGEL,"ENSMUSG00000068196",showPlot = TRUE,trackRegionLabels = setNames(c("name","ID","name"),c("genes","CGI","TFBS")))
```





